pkg/registry/registry.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2018-2021 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package registry
+
+// Registry provides with means for dynamically registering services.
+type Registry interface {
+	// Add registers a Service on the memoryRegistry. Repeated names is allowed, services are distinguished by their metadata.
+	Add(Service) error
+
+	// GetService retrieves a Service and all of its nodes by Service name. It returns []*Service because we can have
+	// multiple versions of the same Service running alongside each others.
+	GetService(string) ([]Service, error)
+}
+
+// Service defines a service.
+type Service interface {
+	GetName() string
+	SetName(string)
```

Comment:
Do we need `SetName`? It is never used. Furthermore I prefer having `Name()` and `Nodes()` as well as passing in a name and a node via Options. See https://golang.org/doc/effective_go#Getters

AFAICT the interface only looks like this because the `memory.Node` implementation needs to have public `Name` and `Nodes` fields in order to unmarshal them using mapstructure. Maybe separating the config struct from the in memory representation makes sense, since we need to use the service names as map keys and `service.Name` fields anyway?